### PR TITLE
Maven: explode final WAR instead of using intermediate exploded WAR directory

### LINKED ISSUE #1091: Explode WAR instead of working on exploded WAR directory for Maven
Explode the WAR file rather than using this directory. The contents of the final WAR may be different from this directory (it's possible to include or exclude files when packaging a WAR. Also the exploded WAR directory is configurable with `<webappDirectory>` and may not be at `build.getFinalName()`.

https://github.com/GoogleContainerTools/jib/pull/1068/files#r221696396

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #1091.

I will have to see if packaging a WAR file in tests works on Windows.

### CODE REVIEW

**Comment by chanseokoh** [comment]
Yeah, as I feared, test code has problems on Windows.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java [hunk 1]
```diff
@@ -187,8 +201,10 @@ public JibContainerBuilder createJibContainerBuilder(
       throws IOException {
     try {
       if (isWarProject()) {
-        Path explodedWarPath =
-            Paths.get(project.getBuild().getDirectory(), project.getBuild().getFinalName());
+        Build build = project.getBuild();
+        Path war = Paths.get(build.getDirectory(), build.getFinalName() + ".war");
```
Comment (TadCordle) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java]: Do we need to check if this file exists?

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java [hunk 2]
```diff
@@ -168,6 +175,32 @@ private static Path getResource(String path) throws URISyntaxException {
...
   }
 
+  private static Path zipUpDirectory(Path sourceRoot, Path targetZip) throws IOException {
+    try (ZipOutputStream zipOut = new ZipOutputStream(Files.newOutputStream(targetZip));
+        Stream<Path> stream = Files.walk(sourceRoot)) {
+      for (Iterator<Path> iterator = stream.iterator(); iterator.hasNext(); ) {
+        Path source = iterator.next();
+        if (source.equals(sourceRoot)) {
+          continue;
+        }
```
Comment (TadCordle) [inline jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java]: `DirectoryWalker` is an option to reduce the code here (we use it elsewhere in the plugins, although it is technically internal to jib-core). It'd be something like

```
for (Path source : new DirectoryWalker(sourceRoot).filterRoot().walk()) {
  ...
}
```

Comment (chanseokoh) [reply jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java]: Good catch. Very nice!

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java (hunk 1 above)
Comment (chanseokoh) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java]: Somehow I couldn't make a comment here. Copying it over from below:

For checking the file, I think we're good. The same happens for the intermediate exploded WAR directory. We catch IOException and suggest to do mvn package.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java [hunk 3]
```diff
@@ -187,8 +201,10 @@ public JibContainerBuilder createJibContainerBuilder(
...
+        Path explodedWarPath = tempDirectoryProvider.newDirectory();
+        ZipUtil.unzip(war, explodedWarPath);
```
Comment (TadCordle) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java]: Sort of unrelated to this PR: I haven't tested, but I noticed that the unzip logic is very similar to the tar extraction logic. I don't really know anything about the zip file format, but I wonder if the fix [here](https://github.com/GoogleContainerTools/jib/pull/2068) should also be applied to `ZipUtil.unzip()`.

Comment (chanseokoh) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java]: I think that is a very good idea.

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java [hunk 4]
```diff
@@ -168,6 +176,29 @@ private static Path getResource(String path) throws URISyntaxException {
     return Paths.get(Resources.getResource(path).toURI());
   }
 
+  private static Path zipUpDirectory(Path sourceRoot, Path targetZip) throws IOException {
+    try (ZipOutputStream zipOut = new ZipOutputStream(Files.newOutputStream(targetZip));
+        Stream<Path> stream = Files.walk(sourceRoot)) {
```
Comment (TadCordle) [inline jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java]: Oh sorry almost forgot, I don't think we need this `Stream<Path> stream = Files.walk(sourceRoot)` anymore.

Comment (chanseokoh) [reply jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java]: Nice catch. Not feeling well today. Should probably go home early.

**Comment by chanseokoh** [comment]
Yeah, I tend to forget that recently. :laughing:

**Pull request merged**